### PR TITLE
Project submission

### LINKED ISSUE #59: add support for Project Submissions
Add support to run project submissions in frameworks like react, node, django etc.

Check [this](https://github.com/coding-blocks/judge-taskmaster/issues/36) issue on `judge-taskmaster` 

- Create a new API endpoint `/api/project`
  - [ ] Create `routes/api/project` directory
  - [ ] Write validator to support following request payload
  ```
  {
    "problem": <problem zip url>,
    "submission": <submission zip url>,
    "lang": <language>,
    "submissionDirs": <directories that can be editted>,
    "timelimit" <timelimit in seconds>,
    "mode": "sync" | "poll" | "callback",
    "callback": <if mode is callback then url>
  }
  ```
  - [ ] Send job to queue required by judge-taskmaster
  - [ ] Implement `onSuccess` listener on the controller

### PR DESCRIPTION
**Pull request by prabalsingh24**
fixes #59

### CODE REVIEW

**Comment by prabalsingh24** [comment]
@jatinkatyal13 how is `ProjectResponse` gonna look like? 
[see this](https://github.com/coding-blocks/judge-taskmaster/blob/78a3cbb99baa3e3e9fddb4cfb6240c0d7dece5e8/src/tasks/scenarios/project.ts#L23) is this how it's gonna look?

**Comment by jatinkatyal13** [comment]
Yes

**Comment by prabalsingh24** [comment]
Have a look at this PR when you're free :)

File: src/routes/api/project/validators.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        problem: Joi
+            .string()
+            .uri()
+            .required(),
+        submission: Joi
+            .string()
+            .uri()
+            .required(),
+        submissionDirs: Joi
+            .array(),
```
Comment (jatinkatyal13) [inline src/routes/api/project/validators.ts:23]: submissionDirs is required

File: src/rabbitmq/jobqueue.ts [hunk 2]
```diff
@@ -26,7 +26,16 @@ export interface RunJob {
...
 }
 
-export type JudgeJob = RunJob | SubmissionJob
+export interface ProjectJob {
+  id: number
+  source: string,
+  lang: string,
+  problem: string,
+  timelimit: number,
+  scenario: string
```
Comment (jatinkatyal13) [inline src/rabbitmq/jobqueue.ts:35]: submissionDirs is also a parameter

File: src/routes/api/project/validators.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        problem: Joi
+            .string()
+            .uri()
+            .required(),
+        submission: Joi
+            .string()
+            .uri()
+            .required(),
+        submissionDirs: Joi
+            .array()
```
Comment (jatinkatyal13) [inline src/routes/api/project/validators.ts:23]: submissionDirs is just a string and NOT an array. It is a glob string and hence named in plural. I get the confusion.

File: src/rabbitmq/jobqueue.ts [hunk 4]
```diff
@@ -50,8 +60,17 @@ amqp.connect(`amqp://${config.AMQP.USER}:${config.AMQP.PASS}@${config.AMQP.HOST}
       jobChannel.consume(successQ, (msg) => {
         debug(`SUCCESS:CONSUME: msg.content = ${msg.content.toString()}`)
 
-        const payload = JSON.parse(msg.content.toString())        
-        const eventName = payload.testcases ? 'submit_result' : 'run_result'
+        const payload = JSON.parse(msg.content.toString())
+        let eventName;
+        if (payload.testcases) {
```
Comment (jatinkatyal13) [inline src/rabbitmq/jobqueue.ts:65]: nitpicking but switch case might be more readable

**Pull request merged**